### ISSUE TITLE
Single algorithms as ReScience publications

### ISSUE THREAD

**Issue by antoine-lizee** [ReScience/ReScience#18]
When working on a large project, researchers need algorithmic tools whose development is not the purpose of the project itself. These tools often don’t get shared properly, because of their small size and/or the fact that they don’t really fit the classic publication scheme (intro-methods-results-discussion). As a result, they’re never really re-used, nor improved on, nor corrected. We believe that these computational blocks, when they have a clear methodological application, could have a place in the new ecosystem of publication ReScience is advocating. Additionally, it would really make sense practically since ReScience is based on github…

Here is an example. @dhimmel , when working on his project (http://thinklab.com/p/rephetio), needed an implementation of the Gerstein-Sonnhammer-Chothia algorithm. Unable to find an existing one, he used the platform of collaborative research [thinklab](http://thinklab.com/discussion/seeking-an-open-source-implementation-of-the-gerstein-sonnhammer-chothia-algorithm/25) to find a suitable collaborator (me) who developed the [tool](https://github.com/antoine-lizee/R-GSC) he needed. After proper testing with example datasets, the tool has been incorporated in the workflow of the bigger project. 
This tool is a well-defined, documented, tested, re-usable piece of code. It should be shared, tested again, compared, improved on, re-used, adapted to other languages; but there is no proper way to **reference** it, since it hasn’t be published. ReScience could bridge this gap.
I would suggest using this example as a case study for our proposition.

Thank you Daniel for the initial proposition! ([tweet](https://twitter.com/dhimmel/status/639126485085868032))

Looking forward to some feedback.

**Comment by ThomasA** [comment]
I am not sure how "big" such contributions must be, but maybe the 
Journal of Open Research Software 
(http://openresearchsoftware.metajnl.com/) could be an option for this 
as well? I have some experience with this journal for a somewhat larger 
software package.
Publishing such algorithms sure is a good idea, but I am thinking that 
maybe we should try not to derail ReScience too much from its original 
definition until it has had time to get properly started and we have 
some more experience with it.

**Comment by ctb** [comment]
On Fri, Sep 11, 2015 at 05:56:52AM -0700, Thomas Arildsen wrote:
> I am not sure how "big" such contributions must be, but maybe the 
> Journal of Open Research Software 
> (http://openresearchsoftware.metajnl.com/) could be an option for this 
> as well? I have some experience with this journal for a somewhat larger 
> software package.
> Publishing such algorithms sure is a good idea, but I am thinking that 
> maybe we should try not to derail ReScience too much from its original 
> definition until it has had time to get properly started and we have 
> some more experience with it.

+1 to JORS, but also +1 to ReScience supporting this.  It could be a super
valuable contribution.  (We have a paper we're submitting to F1000 Research
soon that's basically just this - a decently implemented algorithm that
many people might find useful.  Not big new science, but hecka useful.)

**Comment by rougier** [comment]
I tend to think it fits the journal because the appendix is very much like an article with its own title ("￼A Method to Weight Protein Sequences to Correct for Unequal Representation"). I imagine the original code does not exist anymore ?

Are you also able to reproduce the figures and tables in the appendix ?

**Comment by dhimmel** [comment]
> I imagine the original code does not exist anymore?

I could not find the original code when I looked. Additionally, I couldn't find any readily usable (R or python) implementations.

> Are you also able to reproduce the figures and tables in the appendix?

I think @antoine-lizee [actually found](https://github.com/antoine-lizee/R-GSC/blob/353047286b71d7225e6766d73e7e71c868a6b2c6/test/tests.R#L9) some inaccuracies in the appendix.

**Comment by rougier** [comment]
Ok, then. My question was mostly to know if you can link your implementation to the original article. Finding some inaccuracies is ok and even better if you can point out them in the accompanying article such that the reader knows about the difference.

One last point is whether to use the orignal article title or the one from the appendix to name your own article (+ the [Re]) since it will impact people search when looking for such implementation. But this an be discussed during the review.

**Comment by antoine-lizee** [comment]
Hi all - thanks a lot for your replies.
@ThomasA, thank you for pointing out the Journal of Open Research Software, which didn't catch my attention previously. They offer an interesting platform to publish algorithms, but it seems that they mainly target large, complex pieces of software. It might be a good second option though. And as @ctb said, they don't have to be the only ones providing a publishing service for well-defined pieces of code :-).

@rougier, the original paper dating back from 93', the original implementation of the algorithm - if any - is lost for good. As @dhimmel said, part of my testing is to reproduce their back-of-the-envelope results from the appendix, results which are actually not completely accurate! These inaccuracies could be easily documented (they use low and inconsistent precision in their calculations)

I would go for a more general title than the original one that scopes the algorithm to proteins. It has actually been applied to side-effects of drugs in @dhimmel's project and in the 2008 Science Paper that put @dhimmel on the trail.

This title issue raises the more general question of hosting these single algorithms that don't really fall into the framework of replicated science. This particular example could be presented as the replication of the appendix of the original article - and I'd be happy to submit it as is - but what about, more generally, the single algorithmic blocks I presented above that are just developed among a bigger workflow and never get published? We could mention them as [Re:] of the bigger article, but it's a stretch. Something like [Part:] would be more accurate :-). We might not want to create this complexity too early though, as @ThomasA said. I'll submit my work as [Re:] with a different title, and we'll see what happens, but I'd be interested to know what you guys think.

**Comment by rougier** [comment]
I think we need first to gather more experience on replication before considering extending the scope of the journal. @khinsen, what do you think ?

**Comment by khinsen** [comment]
I agree. We need to keep some focus, and the key term is replication.

I do see the problem though, and it looks quite reasonable to launch another journal like ReScience, borrowing the technical and logistic infrastructure, for publishing reference implementations of algorithms.

There is, however,  another platform with exactly that goal: [Exec&Share](http://www.execandshare.org/). Its particularity is that the published code can be run right on the Exec&Share Web site. However, this also implies restrictions as to what kind of code is acceptable.

**Comment by antoine-lizee** [comment]
@khinsen, thanks for the advice, but it seems that the "particularity" of Exec&Share is a big limitation. Having the platform run the software doesn't really make sense for the vast majority of shareable algorithms I can think of, even though I see how it can be useful, especially to promote the use of the algorithms to a bigger user base.

**Comment by rougier** [comment]
Yes, let's wait for a few months and re-open the issue.
